### ISSUE TITLE
Add ESLint rule to prevent usage of the order CSS property

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#61247]
### Description

See https://github.com/WordPress/gutenberg/issues/61241

Some components come with CSS in JS. 

https://github.com/WordPress/gutenberg/issues/61241 and https://github.com/WordPress/gutenberg/pull/61243 aim to introduce a `stylelint` rule to prevent usage of the `order` CSS property, which is impactful for accessibility.

In the same way, it would be nice to add an ESLint rule to prevent usage of the `order` CSS property in any *.js file.

Unfortunately this isn't exactly my area of expertise so that I'd appreciate some guidance. I did see a similar rule added in https://github.com/WordPress/gutenberg/pull/58130

### Step-by-step reproduction instructions

- Add an `order` CSS property e.g. `order: 1;` to any CSS-in-JS for example in [this file](https://github.com/WordPress/gutenberg/blob/db5e39180b3aadfb965941545c22593fe055e494/packages/components/src/palette-edit/styles.ts)..
-  Run `npm run lint:js`.
- Observe the linter does not report any error.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by afercia** [comment]
I guess the patterns to catch would be, at least, the following ones. With and without leading spaces or tab characters:

```
order: 123;
    order: 123;

order: 123,
    order: 123,

order: '123',
    order: '123',


'order: 123;'
    'order: 123;'
```

**Comment by ghostp13409** [comment]
Hey,  I'm fairly new to open-source and web development, but I would love to give it a try

**Comment by afercia** [comment]
@ghostp13409 welcome. Sure, please do feel free to work on a Pull Request.

I did try to follow a previous example that added another ESLint rule, see https://github.com/WordPress/gutenberg/pull/59022. But I'm not good at regular expressions. Also, I have no idea how that works. I did try to go through the ESLint [custom rules docs](https://eslint.org/docs/latest/extend/custom-rules) and [the docs for selectors](https://eslint.org/docs/latest/extend/selectors) but this isn't really my area of expertise. Wondering if there's any simpler way to detect the `order` CSS property in JS. @mirka any help would be appreciated, when you have a chance 🙇🏽‍♂️ 

It gets complicated because of the several ways CSS can be coded in JS. It could be a key/value pair within an object, or a string, or a concatenated string... etc.

Note the JS linter can be run in the terminal with this command: `npm run lint:js -- --quiet`

**Comment by mirka** [comment]
CSS-in-JS can either be done with strings or objects. The string case can be caught similar to #59022 without much downside, but the object case will be trickier because `{ order: 3 }` is something that could validly appear in code unrelated to styling.

I wouldn't say this rule is particularly necessary at the moment, given that: 

- We currently have zero violations in `@wordpress/components` code, and we generally have more rigorous reviews on CSS that gets merged in the package. (CSS-in-JS is only used in `@wordpress/components`)
- There will be increasingly fewer usages of CSS-in-JS as we migrate off of Emotion.

Might be fine to revisit once we actually encounter an `order` slipping through.

**Comment by Gannu456** [comment]
Hey @afercia , I think i can solve this issue....Can i make a pull request?

**Comment by afercia** [comment]
@Gannu456 thanks for looking into this. Given @mirka's previous comment, I'm not sure it is worth it.

**Comment by cobianzo** [comment]
Hi @afercia . It's been a while since last comment in this issue. and I can't find any PR related to it.
I've checked it carefully, and seems to be a little bit tricky if we want to implement it properly and not to trigger 'false positives' in the linting.
I'm taking it over if you don't mind. cc @ghostp13409 , @Gannu456 

As a comment, your test example proposals 
```
order: 123;
    order: 123;

order: 123,
    order: 123,

order: '123',
    order: '123',


'order: 123;'
    'order: 123;'
```

don't look quite precise to me. I've been digging in the packages codebase, and there are cases where the property 'order' is correctly used in a different context than CSS in JS, therefore we need to ensure that we don't trigger the error in those legit cases.

I'll be pulling a PR today and tomorrow for dicussion, if that's ok with you

**Comment by cobianzo** [comment]
There it goes @afercia , PR ready for reviewing. Since it is my first PR, It would be good if you could check it in a quick look
https://github.com/WordPress/gutenberg/pull/70699 , to see if it looks something similar to what you had in mind.

It was tricky to determine if a developer is actually using the property `order` as a CSS-in-JS, or as anything else. I had to develop some context detection to consider it as styled CSS in JS. 
There were cases like this 
[packages/block-editor/src/store/selectors.js
](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/selectors.js#L738)
where I had to be careful not to trigger a false error.

It uses regular expressions also to detect `order` and not `border` , for example.

I have tested it carefully, confirming that it doesnt trigger false positives, and that it triggers the errors for real cases of `order` CSS injections. The unit test is quite exhaustive, and I also included the documentation about this rule into the 'md' in the same way as the other ones.

The code might look a little overkilling, so I'd appreciate if you could give me a feedback about it.

**Comment by MansurBilalYusufbhai1** [comment]
Hi , I would like to work on this issue. I'm a beginner and will follow the contribution guidelines. Could you please confirm where you'd like the ESLint rule to be added (existing eslint plugin package or a new one? Thanks!

**Comment by mirka** [comment]
Like we said in https://github.com/WordPress/gutenberg/issues/61247#issuecomment-2202936574, this lint rule won't be worth the cost unless the implementation is extremely simple with no false positives. `order` violations slipping through in `@wordpress/components` CSS-in-JS code hasn't been an actual problem yet, so a custom lint rule with any kind of complexity could easily be a net negative if it adds more cost than the cost it saves.

**Comment by cobianzo** [comment]
What about closing this issue @afercia ? We all seem to agree it's not worth the effort. I'm closing my PR too.
